File: templates/_composition/MVVMBasic/Project.SplitView/ViewModels/ShellViewModel.cs
using System;
using System.Linq;
using System.Windows.Input;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Navigation;
using wts.ItemName.Services;
using wts.ItemName.Views;
using wts.ItemName.Helpers;

namespace wts.ItemName.ViewModels
{
    public class ShellViewModel : Observable
    {
        private NavigationView _navigationView;
        private object _selected;
        private ICommand _itemInvokedCommand;

        public object Selected
        {
            get { return _selected; }
            set { Set(ref _selected, value); }
        }

        public ICommand ItemInvokedCommand => _itemInvokedCommand ?? (_itemInvokedCommand = new RelayCommand<NavigationViewItemInvokedEventArgs>(OnItemInvoked));

        public ShellViewModel()
        {
        }

        public void Initialize(Frame frame, NavigationView navigationView)
        {
            _navigationView = navigationView;
            NavigationService.Frame = frame;
            NavigationService.Navigated += Frame_Navigated;
        }

        private void OnItemInvoked(NavigationViewItemInvokedEventArgs args)
        {
            var item = _navigationView.MenuItems
                            .OfType<NavigationViewItem>()
                            .First(menuItem => (string)menuItem.Content == (string)args.InvokedItem);
            var pageType = item.GetValue(NavHelper.NavigateToProperty) as Type;
            NavigationService.Navigate(pageType);
        }

        private void Frame_Navigated(object sender, NavigationEventArgs e)
        {
            var selectedItem = _navigationView.MenuItems
                            .OfType<NavigationViewItem>()
                            .FirstOrDefault(menuItem => IsNavigationViewItemFromPageType(menuItem, e.SourcePageType));

            if (selectedItem != null)
            {
                Selected = selectedItem;
            }
        }

        private bool IsNavigationViewItemFromPageType(NavigationViewItem menuItem, Type sourcePageType)
        {
            var pageType = menuItem.GetValue(NavHelper.NavigateToProperty) as Type;
            return pageType == sourcePageType;
        }
    }
}

File: templates/Uwp/_composition/MVVMLight/Project.SplitViewVB/ViewModels/ShellViewModel.vb
Imports CommonServiceLocator
Imports Windows.UI.Xaml.Controls
Imports Windows.UI.Xaml.Navigation
Imports GalaSoft.MvvmLight
Imports GalaSoft.MvvmLight.Command
Imports wts.ItemName.Services
Imports wts.ItemName.Views
Imports wts.ItemName.Helpers

Namespace ViewModels
    Public Class ShellViewModel
        Inherits ViewModelBase

        Private _navigationView As NavigationView

        Private _selected As NavigationViewItem

        Private _itemInvokedCommand As ICommand

        Public ReadOnly Property NavigationService As NavigationServiceEx
            Get
                Return CommonServiceLocator.ServiceLocator.Current.GetInstance(Of NavigationServiceEx)()
            End Get
        End Property

        Public Property Selected As NavigationViewItem
            Get
                Return _selected
            End Get

            Set(value As NavigationViewItem)
                [Set](_selected, value)
            End Set
        End Property

        Public ReadOnly Property ItemInvokedCommand As ICommand
            Get
                If _itemInvokedCommand Is Nothing Then
                    _itemInvokedCommand = New RelayCommand(Of NavigationViewItemInvokedEventArgs)(AddressOf OnItemInvoked)
                End If

                Return _itemInvokedCommand
            End Get
        End Property

        Public Sub Initialize(frame As Frame, navigationView As NavigationView)
            _navigationView = navigationView
            NavigationService.Frame = frame
            AddHandler NavigationService.Navigated, AddressOf Frame_Navigated
        End Sub

        Private Sub OnItemInvoked(args As NavigationViewItemInvokedEventArgs)
            Dim item = _navigationView.MenuItems.OfType(Of NavigationViewItem)().First(Function(menuItem) CStr(menuItem.Content) = CStr(args.InvokedItem))
            Dim pageKey = TryCast(item.GetValue(NavHelper.NavigateToProperty), String)
            NavigationService.Navigate(pageKey)
        End Sub

        Private Sub Frame_Navigated(sender As Object, e As NavigationEventArgs)
            Selected = _navigationView.MenuItems.OfType(Of NavigationViewItem)().FirstOrDefault(Function(menuItem) IsMenuItemForPageType(menuItem, e.SourcePageType))
        End Sub

        Private Function IsMenuItemForPageType(menuItem As NavigationViewItem, sourcePageType As Type) As Boolean
            Dim navigatedPageKey = NavigationService.GetNameOfRegisteredPage(sourcePageType)
            Dim pageKey = TryCast(menuItem.GetValue(NavHelper.NavigateToProperty), String)
            Return pageKey = navigatedPageKey
        End Function
    End Class
End Namespace
